### PR TITLE
LDGR-404: Implement encrypted saving of private keys to disk

### PR DESCRIPTION
**Pull request by NelisDrost**
Initial implementation, for comment rather than approval. I'll implement some unit tests, and an example if we agree with the general implementation?

### CODE REVIEW

**Comment by NelisDrost** [comment]
Note that I'm getting failures (Chain Code Execution Failure) when running some examples against the ledger - not yet sure whether this is a bug in the ledger/sdk, or compatibility issue.

**Comment by NelisDrost** [comment]
Jenkins failure appears to be an issue importing bcrypt in python 3.5

File: fetchai/ledger/crypto/entity.py [hunk 1]
```diff
@@ -82,17 +97,119 @@ def signing_key(self):
...
+    :param password: plaintext password to use for encryption
+    :param data: plaintext data to encrypt
+    :return: encrypted data, length of original data, initialisation vector for aes, password hashing salt
+    """
+    # Generate hash from password
+    salt = os.urandom(16)
+    hashed_pass = pbkdf2_hmac('sha256', password.encode(), salt, 1000000)
+
+    # Random initialisation vector
+    iv = ''.join([chr(random.randint(33, 127)) for _ in range(16)])
```
Comment (ejfitzgerald) [inline fetchai/ledger/crypto/entity.py]: I suppose that it doesn't really matter, but I guess this is RNG source is not as good as `os.urandom`

File: fetchai/ledger/crypto/entity.py [hunk 2]
```diff
@@ -82,17 +97,119 @@ def signing_key(self):
...
+
+    # Encrypt data using AES
+    aes = pyaes.AESModeOfOperationCBC(hashed_pass, iv=iv)
+
+    # aes = AES.new(hashed_pass, AES.MODE_CBC, iv.encode('ascii'))
```
Comment (ejfitzgerald) [inline fetchai/ledger/crypto/entity.py]: This looks like old code

File: fetchai/ledger/crypto/entity.py [hunk 3]
```diff
@@ -82,17 +97,119 @@ def signing_key(self):
...
+    # Pad data to multiple of 16
+    n = len(data)
+    if n % 16 != 0:
+        data += b' ' * (16 - n % 16)
+
+    encrypted = b''
+    while len(data):
+        encrypted += aes.encrypt(data[:16])
+        data = data[16:]
+    # encrypted = aes.encrypt().encrypt(data)
```
Comment (ejfitzgerald) [inline fetchai/ledger/crypto/entity.py]: Again here

File: setup.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
     author_email='edward.fitzgerald@fetch.ai',
     packages=find_packages(exclude=['contrib', 'docs', 'tests', 'scripts']),
     package_data={'fetchai': ['ledger/parser/etch.grammar']},
-    install_requires=['requests', 'ecdsa', 'msgpack', 'base58', 'lark-parser', 'semver'],
+    install_requires=['requests', 'ecdsa', 'msgpack', 'base58', 'lark-parser', 'semver', 'pyaes'],
```
Comment (ejfitzgerald) [inline setup.py]: Does `pyaes` need any native components?

File: tests/crypto/test_encryption.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        self.assertFalse(entity._strong_password('abc'), "Short password passed")
+        self.assertFalse(entity._strong_password('aaaaaaaaaaaaaa'), "All lower case passed")
+        self.assertFalse(entity._strong_password('11111111111111'), "All numbers passed")
+        self.assertFalse(entity._strong_password('AAAAAAAAAAAAAA'), "All upper case passed")
+        self.assertFalse(entity._strong_password('______________'), "All symbols passed")
+
+        self.assertFalse(entity._strong_password('a1Aaaaaaaaaaaa'), "No symbol passed")
+        self.assertFalse(entity._strong_password('a1a_aaaaaaaaaa'), "No upper case passed")
+        self.assertFalse(entity._strong_password('aaA_aaaaaaaaaa'), "No number passed")
+        self.assertFalse(entity._strong_password('A1_AAAAAAAAAAA'), "No lower case passed")
```
Comment (ejfitzgerald) [inline tests/crypto/test_encryption.py]: 👍

File: fetchai/ledger/crypto/entity.py (hunk 3 above)
Comment (NelisDrost) [reply fetchai/ledger/crypto/entity.py]: Fixed

File: fetchai/ledger/crypto/entity.py (hunk 2 above)
Comment (NelisDrost) [reply fetchai/ledger/crypto/entity.py]: Good catch, fixed.

File: setup.py (hunk 4 above)
Comment (NelisDrost) [reply setup.py]: No, that's why it was chosen over pycryptodome, it's a pure-python implementation of AES.

File: fetchai/ledger/crypto/entity.py (hunk 1 above)
Comment (NelisDrost) [reply fetchai/ledger/crypto/entity.py]: Switched this to os.urandom - there was no good reason not to.

**Pull request merged**